source/material_model/visco_plastic.cc

```diff
@@ -347,14 +375,56 @@ namespace aspect
             out.reaction_terms[i][c] = 0.0;
           // If strain weakening is used, overwrite the first reaction term,
           // which represents the second invariant of the strain tensor
-          if  (use_strain_weakening && this->get_timestep_number() > 0)
+          double edot_ii = 0.;
+          double e_ii = 0.;
+          if  (use_strain_weakening == true && use_finite_strain_tensor == false && this->get_timestep_number() > 0)
             {
-              const double edot_ii = std::max(sqrt(std::fabs(second_invariant(deviator(in.strain_rate[i])))),min_strain_rate);
-
-              // Update reaction term
-              out.reaction_terms[i][0] = edot_ii*this->get_timestep();
+               edot_ii = std::max(sqrt(std::fabs(second_invariant(deviator(in.strain_rate[i])))),min_strain_rate);
+               e_ii = edot_ii*this->get_timestep();
+               // Update reaction term
+              out.reaction_terms[i][0] = e_ii;
             }
         }
+
+      // We need the velocity gradient for the finite strain (they are not included in material model inputs),
+      // so we get them from the finite element.
+      if (in.cell && use_finite_strain_tensor == true && this->get_timestep_number() > 0)
```

Comment:
also test use_strain_weakening is true (or do an assert somewhere that use_finite_strain_tensor can only be true if use_strain_weakening is true)